contracts/defi/provider/CompoundV2Provider.sol

```diff
@@ -0,0 +1,405 @@
+        require(_amount > 0, "Compound: amount cannot be 0");
+        _wallet.invoke(_cToken, 0, abi.encodeWithSignature("redeemUnderlying(uint256)", _amount));
+    }
+
+    /**
+     * @dev Borrows underlying tokens from a cToken contract.
+     * @param _wallet The target wallet.
+     * @param _cToken The cToken contract.
+     * @param _amount The amount of underlying tokens to borrow.
+     */
+    function borrow(BaseWallet _wallet, address _cToken, uint256 _amount) internal {
+        require(_cToken != address(0), "Compound: No market for target token");
+        require(_amount > 0, "Compound: amount cannot be 0");
+        _wallet.invoke(_cToken, 0, abi.encodeWithSignature("borrow(uint256)", _amount));
+    }
+
+    /**
+     * @dev Repays some borrowed underlying tokens to a cToken contract.
+     * @param _wallet The target wallet.
+     * @param _cToken The cToken contract.
+     * @param _amount The amount of underlying to repay.
+     */
+    function repayBorrow(BaseWallet _wallet, address _cToken, uint256 _amount) internal {
+        require(_cToken != address(0), "Compound: No market for target token");
+        require(_amount > 0, "Compound: amount cannot be 0");
+        string memory symbol = CToken(_cToken).symbol();
+        if(keccak256(abi.encodePacked(symbol)) == keccak256(abi.encodePacked("cETH"))) {
+            _wallet.invoke(_cToken, _amount, abi.encodeWithSignature("repayBorrow()"));
+        }
+        else { 
+            address token = CToken(_cToken).underlying();
+            _wallet.invoke(token, 0, abi.encodeWithSignature("approve(address,uint256)", _cToken, _amount));
+            _wallet.invoke(_cToken, 0, abi.encodeWithSignature("repayBorrow(uint256)", _amount));
+        }
+    }
+
+    /**
+     * @dev Enters a cToken market if it was not entered before.
+     * @param _wallet The target wallet.
+     * @param _cToken The cToken contract.
+     * @param _comptroller The comptroller contract.
+     */
+    function enterMarketIfNeeded(BaseWallet _wallet, address _cToken, address _comptroller) internal {
+        bool isEntered = Comptroller(_comptroller).checkMembership(address(_wallet), CToken(_cToken));
+        if(!isEntered) {
+            address[] memory market = new address[](1);
+            market[0] = _cToken;
+            _wallet.invoke(_comptroller, 0, abi.encodeWithSignature("enterMarkets(address[])", market));
+        }
+    }
+
+    /**
+     * @dev Exits a cToken market if there is no more collateral and debt.
+     * @param _wallet The target wallet.
+     * @param _cToken The cToken contract.
+     * @param _comptroller The comptroller contract.
+     */
+    function exitMarketIfNeeded(BaseWallet _wallet, address _cToken, address _comptroller) internal {
+        uint collateral = CToken(_cToken).balanceOf(address(_wallet));
+        uint debt = CToken(_cToken).borrowBalanceStored(address(_wallet));
```

Comment:
I think we should use `borrowBalanceCurrent(address(_wallet))` here instead.

Reply:
In principle yes, but `borrowBalanceCurrent` just accrues all interests first then calls `borrowBalanceStored` which costs a lot of gas. In our case `exitMarketIfNeeded` is always called after an action that already accrues interests so both methods return the same value while `borrowBalanceStored` is much cheaper in terms of gas.